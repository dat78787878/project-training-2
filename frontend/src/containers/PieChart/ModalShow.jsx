import React from 'react';
import { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { getData } from '../../redux/pieChart/actions';
import { Modal, Button } from 'react-bootstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCheck } from '@fortawesome/free-solid-svg-icons';
import { difference } from 'lodash';
const ModalShow = () => {
  const pieData = useSelector((state) => state.pieChart.pieData);
  const { fromDate, toDate } = useSelector((state) => state.dateRange);
  const listAllData = ['Android', 'Windows', 'iOS', 'OsX', 'Unknown', 'Linux'];
  const [listDataModal, setListDataModal] = useState(listAllData);
  const [listDataCheck, setListDataCheck] = useState(pieData.map((val) => val.x));
  const [show, setShow] = useState(false);

  const dispatch = useDispatch();

  const onClickLabel = (val) => {
    setListDataCheck((prevList) => {
      const isChecked = prevList.includes(val);
      if (isChecked) {
        return prevList.filter((check) => check !== val);
      }
      return [...prevList, val];
    });
  };

  const onHide = () => {
    const prevListDataCheck = pieData.map((val) => val.x);
    difference(listDataCheck, prevListDataCheck)
      ? dispatch(getData([fromDate, toDate, listDataCheck]))
      : '';
    setShow(false);
  };

  const handleSearch = (e) => {
    e.preventDefault();
    const keyFilter = e.target.value.trim().toLowerCase();
    if (keyFilter.length > 0) {
      setListDataModal((prev) => {
        return prev.filter((val) => val.toLowerCase().includes(keyFilter));
      });
    } else {
      setListDataModal(listAllData);
    }
  };

  return (
    <div className="modalShow">
      <Button variant="success" className="m-1" onClick={() => setShow(true)}>
        Show modal
      </Button>

      <Modal
        show={show}
        onHide={onHide}
        dialogClassName="modal-custom"
        aria-labelledby="example-custom-modal-styling-title">
        <Modal.Body>
          <input className="modal-custom-input" onChange={handleSearch} />
          {listDataModal.map((val, index) => {
            return (
              <div
                key={index}
                className="d-flex justify-content-between p-2"
                onClick={() => onClickLabel(val)}>
                <span>{val}</span>
                {listDataCheck.includes(val) && <FontAwesomeIcon icon={faCheck} />}
              </div>
            );
          })}
        </Modal.Body>
      </Modal>
    </div>
  );
};

export default ModalShow;
