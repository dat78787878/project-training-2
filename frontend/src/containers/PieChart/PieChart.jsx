import React from 'react';
import { Doughnut } from 'react-chartjs-2';
import { useSelector, useDispatch } from 'react-redux';
import { useEffect } from 'react';
import { getData } from '../../redux/pieChart/actions';
import Error from '../../components/Error/Error';
import { Chart, ArcElement, Tooltip, Legend, Title } from 'chart.js';
import Loading from '../../components/Loading/Loading';

Chart.register(ArcElement, Tooltip, Legend, Title);

const PieChart = () => {
  const { pieData, isLoading, isError } = useSelector((state) => state.pieChart);
  const { fromDate, toDate } = useSelector((state) => state.dateRange);
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(getData([fromDate, toDate, pieData.map((val) => val.x)]));
  }, [fromDate, toDate]);

  const options = {
    elements: {
      arc: {
        borderWidth: 0
      }
    },
    plugins: {
      legend: {
        display: true,
        position: 'right'
      },
      title: {
        display: true,
        text: 'Decive Type',
        align: 'start',
        font: {
          size: 16
        },
        padding: 10
      }
    }
  };

  const data = {
    labels: pieData.map((val) => val.x),
    datasets: [
      {
        data: pieData.map((val) => val.y),
        backgroundColor: ['#00ffc8', '#ea00ff', '#e02b2b', '#2600ff', '#00d9ff', '#00ff0d'],
        hoverBackgroundColor: ['#00ffc8', '#ea00ff', '#e02b2b', '#2600ff', '#00d9ff', '#00ff0d']
      }
    ]
  };

  return (
    <div className="pieChart padding-title">
      {isLoading && <Loading />}
      {isError && <Error />}
      {!isError && !isLoading && <Doughnut data={data} options={options} />}
    </div>
  );
};

export default PieChart;
