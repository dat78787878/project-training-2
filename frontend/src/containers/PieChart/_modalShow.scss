.modal {
  &-content {
    border-radius: 5% !important;
  }

  &-body {
    padding: 0px !important;
    margin: 14px;
  }

  &-custom {
    width: 14% !important;
    position: absolute !important;
    top: 40px;
    left: 50px;
    border: 3px solid #0086c4;
    border-radius: 5%;

    &-input {
      width: 100%;
      border: none ;
      outline: none;
      border-bottom: 3px solid #3cb4ec ;
    }
  }
}
