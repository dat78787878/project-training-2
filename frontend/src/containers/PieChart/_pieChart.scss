.pieChart {
  background-color: white;
  height: 485px;
  position: relative;
  @include pc {
    height: 585px;
  }
  @include tablet {
    height: 710px;
  }
  @include mobile {
    height: 372px;
  }
}
