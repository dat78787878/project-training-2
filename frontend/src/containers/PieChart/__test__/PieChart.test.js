import React from 'react';
import { render, screen } from '@testing-library/react';
import { useSelector, useDispatch } from 'react-redux';
import PieChart from '../PieChart';
window.React = React;

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn()
}));

describe('PieChart Component', () => {
  let mockAppState = {};
  beforeEach(() => {
    useDispatch.mockImplementation(() => () => {});
    useSelector.mockImplementation((callback) => {
      return callback(mockAppState);
    });
  });
  afterEach(() => {
    useSelector.mockClear();
    useDispatch.mockClear();
  });
  it('check loading', () => {
    mockAppState = {
      pieChart: {
        pieData: {},
        isLoading: true,
        isError: false
      }
    };
    const { container } = render(<PieChart />);
    expect(container).toMatchSnapshot();
    expect(container.querySelector('span'));
  });
  it('success task', () => {
    mockAppState = {
      pieChart: {
        pieData: { android: 70, iOS: 30 },
        isLoading: false,
        isError: false
      }
    };
    const { container } = render(<PieChart />);
    expect(container).toMatchSnapshot();
    expect(container.querySelector('canvas')).toBeTruthy();
  });
  it('error task', () => {
    mockAppState = {
      pieChart: {
        pieData: {},
        isLoading: false,
        isError: true
      }
    };
    const { container } = render(<PieChart />);
    expect(container).toMatchSnapshot();
    expect(screen.getByText('Error')).toBeTruthy();
  });
});
