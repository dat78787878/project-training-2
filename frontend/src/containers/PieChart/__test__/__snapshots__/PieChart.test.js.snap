// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChart Component check loading 1`] = `
<div>
  <div
    class="pieChart"
  >
    <span
      class="css-113v409"
    >
      <span
        class="css-e1cvz7"
      />
      <span
        class="css-1p1b5e4"
      />
      <span
        class="css-5j85ug"
      />
      <span
        class="css-1cna3gb"
      />
      <span
        class="css-e4u31e"
      />
      <span
        class="css-30u7o4"
      />
    </span>
  </div>
</div>
`;

exports[`PieChart Component error task 1`] = `
<div>
  <div
    class="pieChart"
  >
    <div
      class="error"
    >
      Error
    </div>
  </div>
</div>
`;

exports[`PieChart Component success task 1`] = `
<div>
  <div
    class="pieChart"
  >
    <canvas
      height="150"
      role="img"
      width="300"
    />
  </div>
</div>
`;
